### PR TITLE
CAMS-461 - WIP - Refactor ACMS migration to use queueTrigger

### PR DESCRIPTION
**Pull request by btposey**
# Purpose

ACMS query was improperly identifying lead cases which was causing us to attempt to migrate very old cases.  

# Major Changes

- Fix lead cases query.
- Convert to using a queue trigger.
- Add admin Function to aid in removing duplicated data.

# Testing/Validation

- Automated tests updated.
- Ran migration in staging, verifying that no failures were encountered and no data was duplicated.

# Notes

The admin Function to delete migrations should not be run after we officially migrate data and go live. After that, we may leave the Function as a stub or have added other utility functionality.

### CODE REVIEW

**Review by fmaddenflx — COMMENTED** [review]
Looking good so far.

File: backend/function-apps/migration/client/acms-migration-trigger.function.ts [hunk 1]
```diff
@@ -32,6 +33,7 @@ export default async function httpStart(
     input: params,
   });
 
+  // TODO: Need to return message like the other HTTP triggers so we can control the HTTP status code.
```
Comment (btposey) [inline backend/function-apps/migration/client/acms-migration-trigger.function.ts:36]: TODO

File: backend/function-apps/migration/orchestration/sub-orchestrator-queue-etl.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+} from '../../../lib/use-cases/acms-orders/acms-orders';
+import { GET_CONSOLIDATIONS } from '../loadConsolidations';
+import { OrchestrationContext } from 'durable-functions';
+
+export function* subOrchestratorETL(context: OrchestrationContext) {
+  const predicateAndPage: AcmsPredicateAndPage = context.df.getInput();
+
+  const leadCaseIds = yield context.df.callActivity(GET_CONSOLIDATIONS, predicateAndPage);
+
+  // TODO: Need to reconsider the report now that we are queuing the ETL tasks. These sane defaults are here to keep the rest of the orchestration from breaking.
```
Comment (btposey) [inline backend/function-apps/migration/orchestration/sub-orchestrator-queue-etl.ts:13]: TODO

File: backend/function-apps/migration/queueTrigger/migrateConsolidations.test.ts [hunk 3]
```diff
@@ -22,25 +26,28 @@ describe('getConsolidations test', () => {
...
     expect(getLeadCaseIdsSpy).toHaveBeenCalledWith(expect.anything(), caseId);
-    expect(actual).toEqual(expected);
   });
 
   test('should properly handle error when getLeadCaseIds controller throws an error', async () => {
     const error = new CamsError('TEST_MODULE', { message: 'getConsolidation Error' });
     jest.spyOn(AcmsOrdersController.prototype, 'migrateConsolidation').mockRejectedValue(error);
 
-    const context = createMockAzureFunctionContext();
+    const context: InvocationContext = {} as InvocationContext;
```
Comment (btposey) [inline backend/function-apps/migration/queueTrigger/migrateConsolidations.test.ts:43]: Maybe? Should be
```suggestion
    const context = createMockAzureFunctionContext();
```

File: backend/lib/use-cases/acms-orders/acms-orders.ts [hunk 4]
```diff
@@ -38,6 +46,7 @@ export type AcmsConsolidation = {
   childCases: AcmsConsolidationChildCase[];
 };
 
+// TODO: These results can be written to a queue to report on.
```
Comment (btposey) [inline backend/lib/use-cases/acms-orders/acms-orders.ts:49]: Done already.
```suggestion
```

File: backend/function-apps/migration/orchestration/sub-orchestrator-queue-etl.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { AcmsPredicate } from '../../../lib/use-cases/acms-orders/acms-orders';
+import { QUEUE_MIGRATION } from '../loadConsolidations';
+import { OrchestrationContext } from 'durable-functions';
+
+export function* subOrchestratorETL(context: OrchestrationContext) {
+  const predicate: AcmsPredicate = context.df.getInput();
+
+  // TODO: Should we log the lead case IDs that are queued??
```
Comment (btposey) [inline backend/function-apps/migration/orchestration/sub-orchestrator-queue-etl.ts:8]: TODO

File: backend/lib/use-cases/acms-orders/acms-orders.ts [hunk 6]
```diff
@@ -42,52 +47,17 @@ export type AcmsTransformationResult = {
...
-  };
-};
-
-export type AcmsPageReport = AcmsAggregate & {
-  predicateAndPage: AcmsPredicateAndPage;
-};
-
-export type AcmsPartitionReport = AcmsAggregate & {
-  predicate: AcmsPredicate;
+  error?: CamsError;
```
Comment (btposey) [inline backend/lib/use-cases/acms-orders/acms-orders.ts:50]: Do we want to add errors to the migration results we store in the "fail" queue??

**Review by fmaddenflx — APPROVED** [review]

File: docs/operations/deployment.md [hunk 7]
```diff
@@ -99,4 +99,30 @@ Note required environment variables and secrets defined in build tool for pipeli
 | AZ_KV_APP_CONFIG_NAME | Secrets | --- | Specifies existing Application Configuration KeyVault |
 | AZ_KV_APP_CONFIG_MANAGED_ID | Secrets | --- | Used by bicep to provide an existing managed identity access the Application Configuration KeyVault Secrets |
 | AZ_KV_APP_CONFIG_RG_NAME | Secrets | --- | Used by bicep to provide scope for the managed identity |
+
+<!-- TODO: FINALIZE THIS -->
```
Comment (jamesobrooks) [inline docs/operations/deployment.md:103]: Let's not forget this TODO.

File: docs/running.md [hunk 8]
```diff
@@ -60,7 +61,7 @@ CAMS_DISABLE_LOCAL_CACHE={true || false}
 
 The API for the CAMS application is implemented with Azure Functions written in Node.js.
 
-Note that any commands listed in this section should be run from the `backend/functions` directory.
+Note that any commands listed in this section should be run from the `backend/` directory.
```
Comment (jamesobrooks) [inline docs/running.md:64]: Typically for notes we use the Docsify syntax `?>`.

File: docs/running.md [hunk 9]
```diff
@@ -121,11 +133,22 @@ MSSQL_DATABASE_DXTR={the name of the DXTR database}
...
 MSSQL_USER={the SQL Server Admin username}
 MSSQL_PASS={the SQL Server Admin user password}
+ACMS_MSSQL_USER={the ACMS SQL Server Admin username}
+ACMS_MSSQL_PASS={the ACMS SQL Server Admin user password}
 # Required for connecting to CAMS SQL server database with managed identity
 MSSQL_CLIENT_ID={OPTIONAL client id of Managed Identity with access}
+ACMS_MSSQL_CLIENT_ID={OPTIONAL client id of Managed Identity with access to the ACMS DB}
+
+AzureWebJobsStorage={Connection String to Azure Functions Storage account}
+MyTaskHub={Durable Functions Task Hub Name}
```
Comment (jamesobrooks) [inline docs/running.md:151]: Just a note that I have to put these in my `local.settings.json`. I'm not sure why it doesn't work in my `.env` file.

File: docs/running.md [hunk 10]
```diff
@@ -166,6 +189,8 @@ installed.
 
 ##### Local Settings File
 
+<!-- TODO: UPDATE THIS IF NEEDED WITH NEW CHANGES -->
```
Comment (jamesobrooks) [inline docs/running.md:192]: Let's not forget this TODO.

File: backend/function-apps/api/admin/admin.function.test.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import {
+  buildTestResponseError,
+  createMockAzureFunctionRequest,
+} from '../../azure/testing-helpers';
+import handler from './admin.function';
+import { UnauthorizedError } from '../../../lib/common-errors/unauthorized-error';
+
+const ADMIN_KEY = 'good-key';
+
+describe.skip('Admin function tests', () => {
```
Comment (jamesobrooks) [inline backend/function-apps/api/admin/admin.function.test.ts:12]: This is temporary while I try something out.

File: backend/function-apps/api/admin/admin.function.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    logger.info(
+      MODULE_NAME,
+      `${request.params?.procedure} procedure triggered by ${JSON.stringify(request.user)}.`,
+    );
+    return toAzureSuccess({
+      body: { data: { procedure: request.params?.procedure, user: request.user?.username } },
+    });
+    // const controller = new AdminController();
+    // const response = await controller.handleRequest(applicationContext);
+    // return toAzureSuccess(response);
```
Comment (jamesobrooks) [inline backend/function-apps/api/admin/admin.function.ts:36]: Temporarily short circuiting to try logging the user info from the portal.

**Review by amorrow-flexion — APPROVED** [review]

**Review by jamesobrooks — APPROVED** [review]

**Pull request merged**